### PR TITLE
GAWB-2499 Added tests of column visibility when downloading metadata

### PR DESCRIPTION
**Pull request by breilly2**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala [hunk 1]
```diff
@@ -332,4 +338,129 @@ class DataSpec extends FreeSpec with WebBrowserSpec
...
+        initialColumns = List("participant_id", "foo", "bar"),
+        defaultShown = Some(List("participant_id", "foo")),
+        defaultHidden = Some(List("bar")),
+        userHidden = Some("foo"),
+        importColumns = Some(List("participant_id", "foo", "bar", "baz")),
+        expectedColumns = List("participant_id", "baz"))
+    }
+  }
+
+  private def testFileDownload(initialColumns: List[String],
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala]: This is about as complex as I'd like to get with a shared test script. Once it starts doing things conditionally depending on function arguments, I tend to want to back off. However, there was enough still shared about the steps and verifications that, to me, the extra logic pays off in readability of the test cases.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/WorkspaceDataPage.scala [hunk 2]
```diff
@@ -18,14 +22,49 @@ class WorkspaceDataPage(namespace: String, name: String)(implicit webDriver: Web
...
+     * automatically close it when the download is complete.
+     */
+    await condition { windowHandles.size == 1 }
+
+    for {
+      path <- downloadPath
+      entityType <- find(CssSelectorQuery(downloadMetadataButton.element.queryString)).get.attribute("data-entity-type")
+    } yield archiveDownloadedFile(s"$path/$entityType.txt")
+
+    def archiveDownloadedFile(sourcePath: String): String = {
```
Comment (UrsaStutsman) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/WorkspaceDataPage.scala]: may want to add comments in here to explain clearly why/what - to be helpful to future you/devs

**Review by UrsaStutsman — APPROVED** [review]
Just the request for some comment in the code - aside from that LGTM!

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala [hunk 3]
```diff
@@ -332,4 +338,129 @@ class DataSpec extends FreeSpec with WebBrowserSpec
...
   }
+
+  "Download should reflect visible columns" - {
+    "no workspace defaults or user preferences" in withWebDriver(downloadPath) { implicit driver =>
+      testFileDownload(
+        initialColumns = List("participant_id", "foo"),
+        expectedColumns = List("participant_id", "foo"))
+    }
+
+    "no workspace defaults, with user preferences" in withWebDriver(downloadPath) { implicit driver =>
```
Comment (andy7i) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala:371]: There's one more test that would be useful to have:
The user can hide the primary key (in this case participant_id) in the UI. When downloaded, the column should still appear. The UI columns and File columns would be different.

**Review by andy7i — APPROVED** [review]
`testFileDownload` is AWESOME. 🌟
Added an idea for a test case. If that's outside the scope of this ticket, we can create a new ticket for it.

**Comment by breilly2** [comment]
@jacmrob, now that this actually works, could you give a quick look at the changes to `hub-compose[-fiab].yml` and `run-tests.sh`?

@UrsaStutsman and @andy7i, there aren't significant changes on the test side since your previous review but you can have a look if you want. I'm torn on taking the time now to add the test Andrew suggested... but I'd really like it to exist so I'm tempted to just make it happen.

@dvoet, I hadn't tagged you on this yet but now that it works it wouldn't be a waste of time to look at it if you're interested.

Successful test run: https://fc-jenkins.dsp-techops.broadinstitute.org/job/fiab-test/1267/

**Review by UrsaStutsman — APPROVED** [review]
Excellent comments

**Comment by breilly2** [comment]
Got verbal 👍  from Jackie.

**Pull request merged**